source/draw/gpu/opengl/OpenGLCanvas.ooc

```diff
@@ -28,6 +28,89 @@ OpenGLCanvas: class extends OpenGLSurface {
 	_target: OpenGLPacked
 	_renderTarget: GLFramebufferObject
 	context ::= this _context as OpenGLContext
+	// TODO: Remove once rendering is done in normalized coordinates
+	_toReference: static func (imageSize: IntVector2D, transform: FloatTransform3D) -> FloatTransform3D {
+		toReference := FloatTransform3D createScaling(imageSize x / 2.0f, imageSize y / 2.0f, 1.0f)
+		toNormalized := FloatTransform3D createScaling(2.0f / imageSize x, 2.0f / imageSize y, 1.0f)
+		toReference * transform * toNormalized
+	}
+	draw: override func ~DrawState (drawState: DrawState) {
+		// Set shader
+		gpuMap: GpuMap = (drawState map == null) ? this context defaultMap : drawState map as GpuMap
+		// Set viewport
+		viewport := (drawState viewportLocal hasZeroArea) ? IntBox2D new(this size) : drawState viewportLocal
+		this context backend setViewport(viewport)
+		// Set transform
+		gpuMap view = _toLocal * _toReference(this size, drawState getTransformNormalized()) * _toLocal
+		// Set model and projection matrices
+		if (this _focalLength > 0.0f) {
+			a := 2.0f * this _focalLength / this size x
+			f := -(this _coordinateTransform e as Float) * 2.0f * this _focalLength / this size y
+			k := (this _farPlane + this _nearPlane) / (this _farPlane - this _nearPlane)
+			o := 2.0f * this _farPlane * this _nearPlane / (this _farPlane - this _nearPlane)
+			gpuMap projection = FloatTransform3D new(a, 0.0f, 0.0f, 0.0f, 0.0f, f, 0.0f, 0.0f, 0.0f, 0.0f, k, -1.0f, 0.0f, 0.0f, o, 0.0f)
+		} else
+			gpuMap projection = FloatTransform3D createScaling(2.0f / this size x, -(this _coordinateTransform e as Float) * 2.0f / this size y, 1.0f)
+		gpuMap model = this _createModelTransform(IntBox2D new(this size))
+		// Set opacity
+		if (drawState opacity < 1.0f)
+			this context backend blend(drawState opacity)
+		else
+			this context backend enableBlend(false)
+		// Set texture
+		if (drawState intputImage)
+			gpuMap add("texture0", drawState intputImage)
+		// Draw
+		gpuMap use()
+		this _bind()
+		this context drawQuad()
+		this _unbind()
+	}
+	draw: override func (action: Func) {
```

Comment:
The below methods are just moved from another class to clean up. This might become a separate PR.